doc/dev/doc/serialization/compact.md

```diff
@@ -0,0 +1,77 @@
+* Can work with no configuration or any kind of factory/serializer registration for .NET classes and structs
+* Platform and language independent
+* Supports partial deserialization of fields, without deserializing the whole objects during queries or indexing
+
+Hazelcast achieves these features by having a well-known schema of objects and replicating them across the cluster which enables members and clients to fetch schemas they don’t have in their local registries. Each serialized object carries just a schema identifier and relies on the schema distribution service or configuration to match identifiers with the actual schema. Once the schemas are fetched, they are cached locally on the members and clients so that the next operations that use the schema do not incur extra costs.
+
+Schemas help Hazelcast to identify the locations of the fields on the serialized binary data. With this information, Hazelcast can deserialize individual fields of the data, without reading the whole binary. This results in a better query and indexing performance.
+
+Schemas can evolve freely by adding or removing fields. Even, the types of the fields can be changed. Multiple versions of the schema may live in the same cluster and both the old and new readers may read the compatible parts of the data. This feature is especially useful in rolling upgrade scenarios.
+
+The Compact serialization does not require any changes in the user classes as it doesn’t need a class to implement a particular interface. Serializers might be implemented and registered separately from the classes.
+
+It also supports zero-configuration use cases by automatically extracting schemas out of the classes and structs records using reflection, which is cached and reused later, with no extra cost.
+
+The underlying format of the compact serialized objects is platform and language independent. Native client supports will be added shortly after promoting this feature to stable status.
+
+## Zero Configuration
+
+Compact serialization can be used without registering a serializer for a type. Hazelcast will then try to extract a schema out of the class, using reflection, by inspecting all public properties. If successful, it registers the reflection-based serializer associated with the extracted schema and uses it while serializing and deserializing instances of that class. If the automatic schema extraction fails, Hazelcast throws an exception.
+
+Currently, most primitive types (`bool`, `int`...) are supported, as well as enums, arrays of those types, and nested classes. More advanced classes (e.g. `List<T>`) are not supported yet. Intefaces are not supported.
+
+## Compact Serializer
+
+Another way to use compact serialization is to implement the `ICompactSerializer<T>` interface for a type, and register it in the configuration. A basic serializer could look like:
+
+```csharp
+public class EmployeeSerializer : ICompactSerializer<Employee>
```

Comment:
@zpqrtbnk 

Don't we have `getClassName()` in .net client? Why?

Reply:
This was discussed with Metin. In .NET types are either open generics, or closed generics. For instance, `List<>` is an open generic and more or less corresponds to the Java type (where types are erased). And `List<int>` is a closed generic. If we have a `ICompactSerializer<Thing<Foo>>` then `serializer.GetType()` is going to return `Thing<Foo>` as a closed generic, the type is not erased, we can fully determine the serialized type - something you cannot do in Java. So... .NET does not require a `getClassName()` method because we can obtain the complete details from the type - such a method would be redundant and error-prone.